docs: api (#332)
mend/renovate-ce-ee: docs/api.md

# Renovate API

> Renovate CE/EE exposes a REST API that can be used to interact programmatically with Renovate.

Reviewer: ```suggestion
Renovate CE/EE exposes a REST API that you can use to interact programmatically with Renovate.
```

## Enabling and Authentication

The API can be enabled by setting the `MEND_RNV_ADMIN_API_ENABLED` environment variable to `true`.
You must also configure an API secret by setting the `MEND_RNV_SERVER_API_SECRET` variable.

Authentication is done via HTTP Auth, using the API secret as the password.
For example if the secret is `renovateapi` then you would authenticate with:

```

>   Authorization: renovateapi

Reviewer 2: Would this not be 
```suggestion
  Authorization: Bearer renovateapi
```

Or

```suggestion
  Authorization: Token renovateapi
```

Or something similar? Some HTTP clients may expect there to be [a scheme in the authorization header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization)

Author: I _thought_ it should have such a prefix, but the internal testing docs we had indicated otherwise, so this is the primary reason I have paused for review

Reviewer 3: no scheme is needed

Author: It would be nice if `Bearer` is optional as people may assume it's needed, for now I'll resolve and merge this

```

## Endpoints

### Health

`GET /health`

Returns 200 if API is healthy.

### Queue status

`GET /api/job/queue`

Returns the current status of the job queue, including number of pending jobs.

> `GET /api/task/queue`

Reviewer 2: Nice, this is what used to be available from I think the `/status` endpoint on Renovate On Prem, right? Glad to see it's now behind an authN check 👏🏽 (my org was concerned that it may expose information about private repo names to the public Internet, so we restricted traffic at the Load Balancer)

Author: Yes, that was the intention. I don't think we exposed "task" information as that's a new internal concept part of the CE rewrite, but we exposed _job_ information

Reviewer 3: Health endpoint is not authenticated and responds with a status code only.
All other endpoints are authenticated tho

Returns the current status of the task queue, including number of pending tasks.

> Generally speaking, tasks are internal implementation details (such as sync'ing) so this API is exposed primarily for the purpose of troubleshooting if something is going wrong.

Reviewer: ```suggestion
Generally speaking, tasks are internal implementation details, such as syncing.
As end-user you usually do not need to worry about tasks.
This API is exposed primarily to help you troubleshoot if something is going wrong.
```

We should split up the long sentence into two (or three) parts:

- Tasks is internal implementation detail
- API is for troubleshooting
- _Optional:_ explain that you usually won't need to worry about tasks at all

### Sync and jobs

`POST /api/sync`

Triggers an immediate repository sync against the platform/server.

> Should not normally be necessary but useful if you think Renovate's internal state may have become out of sync (e.g. a new repo is missing).

Reviewer: ```suggestion
Normally you don't need this endpoint.
But it can be useful if you think Renovate's internal state has become out of sync: for example when a new repository is missing.
```

`POST /api/job/add`

> This endpoint allows adding a new job to the queue. The request body should contain a single repository:

Reviewer: ```suggestion
This endpoint allows adding a new job to the queue.
The request body must contain a single repository:
```

- One sentence per line
- Do we mean _should_ or _must_ in the second sentence?

```
{ "repository": "some-org/some-repo" }

> ```

Reviewer: This looks like a JSON object, should it use the JSON styling for the codeblock?

Author: Yes!

Authentication is done via HTTP Auth, using the API secret as the password.
For example if the secret is `renovateapi` then you would authenticate with:

```
  Authorization: renovateapi
```

## Endpoints

### Health

`GET /health`

Returns 200 if API is healthy.

### Queue status

`GET /api/job/queue`

Returns the current status of the job queue, including number of pending jobs.

`GET /api/task/queue`

Returns the current status of the task queue, including number of pending tasks.
Generally speaking, tasks are internal implementation details, such as syncing.
As end-user you usually do not need to worry about tasks.
This API is exposed primarily to help you troubleshoot if something is going wrong.

### Sync and jobs

`POST /api/sync`

Triggers an immediate repository sync against the platform/server.
Normally you don't need this endpoint.
But it can be useful if you think Renovate's internal state has become out of sync: for example when a new repository is missing.

`POST /api/job/add`

This endpoint allows adding a new job to the queue.
The request body must contain a single repository:

> ```

Reviewer: ```suggestion
```json
```
